### PR TITLE
net: lwm2m: Software management object

### PR DESCRIPTION
**Pull request by jarlamsa**
Initial implementation for software management object; Object number 9 in the LwM2M -specification.
http://www.openmobilealliance.org/release/LWM2M_SWMGMT/V1_0_2-20210119-A/OMA-TS-LWM2M_SwMgmt-V1_0_2-20210119-A.pdf

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_obj_device.c [hunk 1]
```diff
@@ -132,7 +132,7 @@ static struct lwm2m_engine_res_inst *error_code_ri;
 /* callbacks */
 
 static int reset_error_list_cb(uint16_t obj_inst_id,
-			       uint8_t *args, uint16_t args_len)
+				   uint8_t *args, uint16_t args_len)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_device.c]: I don't think this change was intended.

File: subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+	}
+
+	return 0;
+}
+
+static int package_uri_write_cb(uint16_t obj_inst_id, uint16_t res_id,
+				uint16_t res_inst_id, uint8_t *data, uint16_t data_len,
+				bool last_block, size_t total_size)
+{
+	LOG_ERR("Not supported");
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c:466]: What are the plans for PULL mode support?

File: subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+
+static int lwm2m_software_mgmt_init(const struct device *dev)
+{
+	struct lwm2m_engine_obj_inst *obj_inst = NULL;
+	int ret = 0;
+
+	software_mgmt.obj_id = LWM2M_OBJECT_SOFTWARE_MANAGEMENT_ID;
+	software_mgmt.version_major = SOFTWARE_MGMT_VERSION_MAJOR;
+	software_mgmt.version_minor = SOFTWARE_MGMT_VERSION_MINOR;
+	software_mgmt.is_core = true;
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c:595]: This doesn't seem right, the SW management object is not defined by the core spec.

File: subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+	software_mgmt.field_count = ARRAY_SIZE(fields);
+	software_mgmt.max_instance_count = MAX_INSTANCE_COUNT;
+	software_mgmt.create_cb = software_mgmt_create;
+	lwm2m_register_obj(&software_mgmt);
+
+	/* auto create the first instance */
+	ret = lwm2m_create_obj_inst(LWM2M_OBJECT_SOFTWARE_MANAGEMENT_ID, 0, &obj_inst);
+	if (ret < 0) {
+		LOG_ERR("Create LWM2M server instance 0 error: %d", ret);
+	}
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c:606]: I just wonder, does it make sense to create an instance by default? Perhps it'd be better to leave it up to the user to define what software/firmware components can be updated?

**Review by rlubos — COMMENTED** [review]
Initial high level overview.

File: subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+    OBJ_FIELD_DATA(SOFTWARE_MGMT_SOFTWARE_COMPONENT_TREE_LENGTH_ID, R_OPT, U8)
+};
+
+static struct lwm2m_engine_obj_inst inst[MAX_INSTANCE_COUNT];
+static struct lwm2m_engine_res res[MAX_INSTANCE_COUNT][SOFTWARE_MGMT_MAX_ID];
+static struct lwm2m_engine_res_inst res_inst[MAX_INSTANCE_COUNT][RESOURCE_INSTANCE_COUNT];
+
+static uint8_t lwm2m_software_mgmt_instance_id_to_index(uint16_t obj_inst_id)
+{
+    int i;
```
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c]: Scratch that, already fixed.

**Comment by carlescufi** [comment]
Closing in favor of https://github.com/zephyrproject-rtos/zephyr/pull/40552

**Pull request closed without merging**